ArduCopter/takeoff.cpp

```diff
@@ -85,16 +80,32 @@ void Mode::_TakeOff::do_pilot_takeoff(float& pilot_climb_rate_cm)
         return;
     }
 
-    float pos_z = take_off_complete_alt;
-    float vel_z = pilot_climb_rate_cm;
+    if (copter.ap.land_complete) {
+        // send throttle to attitude controller with angle boost
+        float throttle = constrain_float(copter.attitude_control->get_throttle_in() + copter.G_Dt / copter.g2.takeoff_throttle_slew_time, 0.0, 1.0);
+        copter.attitude_control->set_throttle_out(throttle, true, 0.0);
+        // tell position controller to reset alt target and reset I terms
+        copter.flightmode->set_throttle_takeoff();
+        if (throttle >= 0.9 || 
+            (copter.pos_control->get_z_accel_cmss() >= 0.5 * copter.pos_control->get_max_accel_z_cmss()) ||
+            (copter.pos_control->get_vel_desired_cms().z >= constrain_float(pilot_climb_rate_cm, copter.pos_control->get_max_speed_up_cms() * 0.1, copter.pos_control->get_max_speed_up_cms() * 0.5)) || 
+            (is_positive(take_off_complete_alt) && copter.pos_control->get_pos_target_z_cm() - take_off_start_alt > 0.5 * (take_off_complete_alt  - take_off_start_alt))) {
+            copter.set_land_complete(false);
```

Comment:
We need to remember that the takeoff can be interrupted if the user switches modes during takeoff.  I worry that this could lead to this bit of takeoff detection code not running and leaving the vehicle with land_complete = false.